File: app/src/main/java/de/bentolor/ahamanager/AhaLab.java
package de.bentolor.ahamanager;

import android.content.Context;
import android.util.Log;

import java.util.ArrayList;
import java.util.UUID;

public class AhaLab {
    private static final String TAG = "AhaLab";
    private static final String FILENAME = "ahas.json";
    private static AhaLab sAhaLab;
    private ArrayList<Aha> mAhas;
    private AhaJSONSerializer mSerializer;

    private AhaLab(Context appContext) {
        mSerializer = new AhaJSONSerializer(appContext, FILENAME);

        try {
            mAhas = mSerializer.loadCrimes();
        } catch (Exception e) {
            mAhas = new ArrayList<Aha>();
            Log.e(TAG, "Error loading Ahas: ", e);
        }

        createAhas(7);
    }

    public static AhaLab get(Context c) {
        if (sAhaLab == null) {
            sAhaLab = new AhaLab(c.getApplicationContext());
        }
        return sAhaLab;
    }

    public ArrayList<Aha> getAhas() {
        return mAhas;
    }

    public Aha getAha(UUID id) {
        for (Aha c : mAhas) {
            if (c.getId().equals(id))
                return c;
        }
        return null;
    }

    public void addAha(Aha c) {
        mAhas.add(c);
    }

    public void deleteAha(Aha c) {
        mAhas.remove(c);
    }

    public boolean saveAhas() {
        try {
            mSerializer.saveAhas(mAhas);
            Log.d(TAG, "Ahas saved to file");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Error saving Ahas: ", e);
            return false;
        }
    }

    private void createAhas(int count) {
        for (int i = 0; i < count; i++) {
            Aha aha = new Aha();
            aha.setTitle("Aha #" + i);
            aha.setUseful((i % 2) == 0); // Alternate
            mAhas.add(aha);
        }
    }
}

File: app/src/main/kotlin/de/bentolor/katamanager/AhaLab.kt
package de.bentolor.katamanager

import android.content.Context
import android.util.Log
import java.util.*

class AhaLab private constructor(appContext: Context) {

    var ahas: MutableList<Aha>
        private set

    private val mSerializer: AhaJSONSerializer = AhaJSONSerializer(appContext, FILENAME)

    init {
        try {
            ahas = mSerializer.loadCrimes()
            Log.d(TAG, "${ahas.size} Ahas loaded")
        } catch (e: Exception) {
            ahas = mutableListOf()
            Log.e(TAG, "Error loading Ahas: ", e)
        }

        if (ahas.size == 0) createAhas(7)
    }

    fun getAha(id: UUID): Aha? = ahas.filter { it.id == id }.firstOrNull()

    fun addAha(c: Aha) = ahas.add(c)

    fun deleteAha(c: Aha) = ahas.remove(c)

    fun saveAhas(): Boolean {
        try {
            mSerializer.saveAhas(ahas)
            Log.d(TAG, "Ahas saved to file")
            return true
        } catch (e: Exception) {
            Log.e(TAG, "Error saving Ahas: ", e)
            return false
        }
    }

    private fun createAhas(count: Int) {
        for (i in 0..count - 1) {
            val aha = Aha(title = "Kata #$i", isUseful = (i % 2 == 0))
            ahas.add(aha)
        }
    }

    companion object {
        private val TAG = "AhaLab"
        private val FILENAME = "ahas.json"
        private var sAhaLab: AhaLab? = null

        operator fun get(c: Context): AhaLab {
            val ahaLab = sAhaLab ?: AhaLab(c.applicationContext)
            sAhaLab = ahaLab
            return ahaLab
        }
    }
}
